pkg/scheduler/core/generic_scheduler.go

```diff
@@ -229,31 +224,133 @@ func (g *genericScheduler) Preempt(pod *v1.Pod, nodeLister algorithm.NodeLister,
+					nodeToVictims, g.cachedNodeInfoMap, pdbs); err != nil {
+					return "", err
+				}
+				// If node list has all removed, no preemption will happen, skip other extenders.
+				if len(nodeToVictims) == 0 {
+					break
+				}
+			} else {
+				nonPreemptingExtenders = append(nonPreemptingExtenders, extender)
+			}
+		}
+	}
+
+	// Unconditional call pickOneNodePassingNonPreemptingExtenders() since if nonPreemptingExtenders is empty,
+	// this will return result of pickOneNodeForPreemption() which is expected.
+	candidateNodeName = g.pickOneNodePassingNonPreemptingExtenders(pod, nodeToVictims, nonPreemptingExtenders)
+
+	return candidateNodeName, nil
+}
+
+// pickOneNodePassingNonPreemptingExtenders checks nodeToVictims with extenders which do not support ProcessPreemption(),
+// and returns the candidate node after checking.
+func (g *genericScheduler) pickOneNodePassingNonPreemptingExtenders(
+	pod *v1.Pod,
+	nodeToVictims map[string]*schedulerapi.Victims,
+	nonPreemptingExtenders []algorithm.SchedulerExtender) string {
 	for len(nodeToVictims) > 0 {
-		node := pickOneNodeForPreemption(nodeToVictims)
-		if node == nil {
-			return nil, nil, nil, err
+		// 1. Pick a node from potential preemption node list
+		nodeName := pickOneNodeForPreemption(nodeToVictims)
+		if len(nodeName) == 0 {
+			return ""
 		}
-		passes, pErr := nodePassesExtendersForPreemption(pod, node.Name, nodeToVictims[node].pods, g.cachedNodeInfoMap, g.extenders)
+		// 2. Check if it passes Filter() check of all nonPreemptingExtenders
+		passes, pErr := nodePassesNonPreemptingExtenders(pod, nodeName, nodeToVictims[nodeName].Pods,
+			g.cachedNodeInfoMap, nonPreemptingExtenders)
+		// 3. If passes, we can use this node as preemption candidate.
 		if passes && pErr == nil {
-			// Lower priority pods nominated to run on this node, may no longer fit on
-			// this node. So, we should remove their nomination. Removing their
-			// nomination updates these pods and moves them to the active queue. It
-			// lets scheduler find another place for them.
-			nominatedPods := g.getLowerPriorityNominatedPods(pod, node.Name)
-			return node, nodeToVictims[node].pods, nominatedPods, err
+			return nodeName
 		}
 		if pErr != nil {
-			glog.Errorf("Error occurred while checking extenders for preemption on node %v: %v", node, pErr)
+			glog.Errorf("Error occurred while checking extenders for preemption on node %v: %v",
+				nodeName, pErr)
 		}
-		// Remove the node from the map and try to pick a different node.
-		delete(nodeToVictims, node)
+		// 4. Otherwise, remove the node from the map and try to pick a different node.
+		delete(nodeToVictims, nodeName)
 	}
-	return nil, nil, nil, err
+	return ""
+}
+
+// nodePassesNonPreemptingExtenders checks node with extenders' Filter() to make sure this node
+// is possible to be preempted.
+// This is a fall back method for those extenders that do not support preemption.
+func nodePassesNonPreemptingExtenders(
```

Comment:
I am sure why this is needed. If Filter passes here, it would have passed earlier before hitting the Preemption phase. Can we assume if someone is deploying an extender without preemption support, they wouldn't enable preemption in the scheduler?

Reply:
> If Filter passes here, it would have passed earlier before hitting the Preemption phase

Emm, I am not sure I understand your concern. What `nodePassesNonPreemptingExtenders` is checking is candidate nodes for preemption, that means they should have failed during predicate phase.

Reply:
Got it, thanks!